api/src/main/java/org/openmrs/module/fhir2/providers/r3/DiagnosticReportFhirResourceProvider.java

```diff
@@ -68,4 +75,22 @@ public MethodOutcome updateDiagnosticReport(@IdParam IdType id, @ResourceParam D
 		return FhirProviderUtils.buildUpdate(
 		    diagnosticReportService.update(idPart, DiagnosticReport30_40.convertDiagnosticReport(diagnosticReport)));
 	}
+	
+	@Search
+	public IBundleProvider searchForDiagnosticReports(
+	        @OptionalParam(name = DiagnosticReport.SP_ENCOUNTER, chainWhitelist = {
+	                "" }, targetTypes = Encounter.class) ReferenceAndListParam encounterReference,
+	        @OptionalParam(name = DiagnosticReport.SP_PATIENT, chainWhitelist = { "", Patient.SP_IDENTIFIER,
+	                Patient.SP_GIVEN, Patient.SP_FAMILY,
+	                Patient.SP_NAME }, targetTypes = Patient.class) ReferenceAndListParam patientReference,
+	        @OptionalParam(name = DiagnosticReport.SP_SUBJECT, chainWhitelist = { "", Patient.SP_IDENTIFIER, Patient.SP_NAME,
+	                Patient.SP_GIVEN, Patient.SP_FAMILY }) ReferenceAndListParam subjectReference,
+	        @OptionalParam(name = DiagnosticReport.SP_ISSUED) DateRangeParam issueDate,
+	        @OptionalParam(name = DiagnosticReport.SP_CODE) TokenAndListParam code, @Sort SortSpec sort) {
+		if (patientReference == null) {
+			patientReference = subjectReference;
+		}
+		return diagnosticReportService.searchForDiagnosticReports(encounterReference, patientReference, issueDate, code,
+		    sort);
```

Comment:
One thing I had completely forgotten is conversion to R3. You see the code above will always return FHIR R4 resources. The service layer calls `searchQuery.getQueryResults(theParams, dao, translator);` and goes down to here;
```java
@Override
	public List<IBaseResource> getResources(int fromIndex, int toIndex) {
		int firstResult = 0;
		if (fromIndex >= 0) {
			firstResult = fromIndex;
		}
		
		int maxResults = -1;
		if (toIndex - fromIndex > 0) {
			maxResults = toIndex - fromIndex;
		}
		
		return dao.search(theParams, firstResult, maxResults).stream().map(translator::toFhirResource)
		        .collect(Collectors.toList());
	}
```
where it's only been translated to R4 (no conversions).

Initially, this `Bundle30_40.convertBundle()` was been used to convert to R3 which doesn't for IBundeProvider. We need to find a way of converting IBundleProvider resources to R3.

My Initial thought was that conversions should go down to the translators' layer where we add a new method 
`org.hl7.fhir.dstu3.model.Patient toFhirR3(org.openmrs.Patient patient)` 
then implementation
`Patient30_40.convertPatient(toFhirResource(patient))`
If the conversion is not satisfactory, this will give us the flexibility of doing them manually.

cc: @ibacher

Reply:
@corneliouzbett At a first pass, I think I'd want to keep the concept of "version conversions" separate from the concept of "translations". The rationale for this being that FHIR resources have been renamed or changed across versions. So, for instance, `Person` is pretty much consistently a FHIR resource from the initial draft of what became FHIR until the present, but, e.g., both ProcedureRequest (R3) and ReferralRequest (R3) merged into ServiceRequest (R4).

At a guess, one way to implement this is via a `VersionConverter` class that is supplied in the R3 version of the Provider and omitted in the R4 version.

Reply:
> @corneliouzbett At a first pass, I think I'd want to keep the concept of "version conversions" separate from the concept of "translations". The rationale for this being that FHIR resources have been renamed or changed across versions. So, for instance, `Person` is pretty much consistently a FHIR resource from the initial draft of what became FHIR until the present, but, e.g., both ProcedureRequest (R3) and ReferralRequest (R3) merged into ServiceRequest (R4).
> 
> At a guess, one way to implement this is via a `VersionConverter` class that is supplied in the R3 version of the Provider and omitted in the R4 version.

That's okay. my point is there's no implementation or a method for converting R4 IBundleProvider to R3 & vice versa. So all R3 providers return paged R4 resources except for GET methods.

Reply:
Yeah, we need to fix that.